API4: Allow save() to match on null values

highfalutin opened the pull request:
Before
----------------------------------------
Code was trying to allow save() to match on null values, but was shooting itself in the foot.

After
----------------------------------------
Matching on fields works even when their value is null.

highfalutin committed:
remove contradictory isset()

colemanw commented:
> Code was trying to allow save() to match on null values

I'm not entirely sure that it was. But I'm not against updating it to do so, with a unit test, and I think we ought to add an additional safety-guard that a positive match must include at least one **non**-null value.

highfalutin committed:
don't allow all match values to be empty

highfalutin committed:
allow all match values to be empty

highfalutin committed:
Test that we get a match on NULL

highfalutin commented:
@colemanw, the already-merged PR #22882 (@mattwire) intended to allow matching on null -- see the [comment](https://github.com/civicrm/civicrm-core/pull/22882/files#diff-ccb444858b7cfcd34c34786bd76003cdccad4c53fbadef2b89df5ba7a94e82f2R146). I'm keen on having this work properly.

I added a test. I also added in the "safety guard" ([this commit](https://github.com/civicrm/civicrm-core/pull/24971/commits/ea34f97f03f9885877010d08809df508f1c11cc1)) but after thinking about it, I took it back out. I don't know why we'd assume people *won't* want to match on a set of empty values. The function already aborts if the number of matches > 1, and that seems like enough of a safety guard.

colemanw commented:
Wow, that was fast!
Ok, to clarify (the test is a little hard to grok), is NULL being treated like a wildcard that matches any value? Or does NULL only match NULL?

highfalutin committed:
make test easier to understand and review

highfalutin commented:
NULL or empty-string only matches NULL or empty-string (using the IS EMPTY operator). And I agree, the test was impressively clever, but almost impossible to read, so I've rewritten it from scratch.

highfalutin committed:
correct typo in comment

colemanw commented on tests/phpunit/api/v4/Action/SaveTest.php:
```diff
@@ -29,79 +28,93 @@
+      Contact::delete(FALSE)
+        ->setUseTrash(FALSE)
+        ->addWhere('id', 'IN', $allContactIds)
+        ->execute();
```
It was thoughtful of you to add this cleanup at the end, but that's not how we do test cleanup :)
For one thing, it's a pain for test authors to remember to do this every time, and, more importantly, if the test fails then it will never get to this step!

Best practice these days is to use `$this->createTestRecord()` instead of the api. As the name implies, the function creates test records... but it also automatically cleans them up afterwards, even if the test crashes halfway through!

colemanw commented:
@highfalutin test is fine now can you just remove the bit at the end?

colemanw commented:
@highfalutin are you able to update this? It would be great to get it merged.

highfalutin commented:
@colemanw a busy few weeks here but I'll get to it. IIRC, we can't take out the bit at the end and still have a passing test, but I'll take another look

highfalutin committed:
use data provider instead of foreach loop

highfalutin commented:
@colemanw I got rid of the "delete" action, and switched to relying on TransactionalInterface to handle cleanup. Back to using a data provider. But hopefully the test is still not too hard to grok.

highfalutin commented:
I believe the test failure is unrelated.

civicrm-builder commented:
Can one of the admins verify this patch?

highfalutin commented:
civibot, test this please

highfalutin commented:
@colemanw can we get this into the new rc?

colemanw commented:
@civicrm-builder retest this please

colemanw commented:
Yes, thanks for your work on this @highfalutin
